src/server/scripts/Kalimdor/RagefireChasm/boss_lava_guard_gordoth.cpp

```diff
@@ -0,0 +1,221 @@
+    }
+
+    void JustEngagedWith(Unit* who) override
+    {
+        BossAI::JustEngagedWith(who);
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, me, 1);
+        events.ScheduleEvent(EVENT_GROUND_RUPTURE, 7300ms);
+        events.ScheduleEvent(EVENT_SEISMIC_SLAM, 13300ms);
+    }
+
+    void DamageTaken(Unit* /*attacker*/, uint32& damage, DamageEffectType /*damageType*/, SpellInfo const* /*spellInfo = nullptr*/) override
+    {
+        if (me->HealthBelowPctDamaged(31, damage))
+            DoCastSelf(SPELL_ENRAGE);
+    }
+
+    void DoAction(int32 action) override
+    {
+        switch (action)
+        {
+           case ACTION_JUMP:
+           {
+                me->SetAIAnimKitId(ANIMKIT_GORDOTH_NONE);
+                DoCastSelf(SPELL_JAIL_BREAK);
+                me->GetMotionMaster()->MoveJumpWithGravity(GordothJumpPoint, 50.0f, 55.5477f, POINT_JUMP);
+
+                scheduler.Schedule(30ms, [this](TaskContext /*task*/)
+                {
+                    if (GameObject* go = GetClosestGameObjectWithEntry(me, GORDOTH_CAGE, 50.0f))
+                        go->SetGoState(GO_STATE_ACTIVE);
+                });
+                break;
+            }
+            case ACTION_KILL_RESEARCHERS:
+            {
+                DoCast(SPELL_GROUND_SLAM);
+                me->SetFacingTo(0.244346f);
+                me->SetHomePosition(GordothJumpPoint);
+
+                std::vector<Creature*> darkShamanResearcherList;
+                GetCreatureListWithEntryInGrid(darkShamanResearcherList, me, NPC_DARK_SHAMAN_RESEARCHER, 25.0f);
+                for (Creature* darkShamanResearcher : darkShamanResearcherList)
+                {
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_1, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_1, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_2, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_2, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_3, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_3, TRIGGERED_FULL_MASK);
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_MEAT_EXPLOSION_3, TRIGGERED_FULL_MASK); // According to sniff
+                    darkShamanResearcher->CastSpell(darkShamanResearcher, SPELL_BLOODY_SUICIDE, false);
+                }
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    void MovementInform(uint32 type, uint32 pointId) override
+    {
+        if (type != EFFECT_MOTION_TYPE)
+            return;
+
+        if (pointId == POINT_JUMP)
+            DoAction(ACTION_KILL_RESEARCHERS);
```

Comment:
you can move the behavior of the action directly here. `DoAction` shouldn't be used to call a method within the same script :D

its mostly used if you want to trigger something on a different script. E.g. npc A triggers something on npc B